File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root: Optional[TreeNode], lvl: int) -> int:
        if not root:
            return 0
        left = self.solve(root.left, lvl + 1)
        right = self.solve(root.right, lvl + 1)
        if right == left:
            if left + lvl >= self.depth:
                self.res = root
                self.depth = left + lvl
        return max(left, right) + 1
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.res = None
        self.depth = -1
        self.solve(root, 0)
        return self.res
        

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        TreeNode* res = nullptr;
        int depth = -1;
        solve(root, 0, depth, res);
        return res;
    }
private:
    int solve(TreeNode* root, int level, int& depth, TreeNode*& res)
    {
        if(!root)
        {
            return 0;
        }
        int left = solve(root->left, level + 1, depth, res);
        int right = solve(root->right, level + 1, depth, res);
        if(left == right)
        {
            if(left + level >= depth)
            {
                res = root;
                depth = left + level;
            }
        }
        return max(left, right) + 1;
    }
};